File: system-test/java/src/test/java/com/mycompany/myshop/systemtest/latest/acceptance/PlaceOrderPositiveTest.java
package com.mycompany.myshop.systemtest.latest.acceptance;

import com.mycompany.myshop.systemtest.latest.acceptance.base.BaseAcceptanceTest;
import com.mycompany.myshop.testkit.channel.ChannelType;
import com.mycompany.myshop.testkit.driver.port.dtos.OrderStatus;
import com.optivem.testing.Channel;
import com.optivem.testing.DataSource;
import org.junit.jupiter.api.TestTemplate;

class PlaceOrderPositiveTest extends BaseAcceptanceTest {
    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void shouldBeAbleToPlaceOrderForValidInput() {
        scenario
                .given().product()
                    .withSku("ABC")
                    .withUnitPrice(20.00)
                .and().country()
                    .withCode("US")
                    .withTaxRate(0.10)
                .when().placeOrder()
                    .withSku("ABC")
                    .withQuantity(5)
                    .withCountry("US")
                .then().shouldSucceed();
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void orderStatusShouldBePlacedAfterPlacingOrder() {
        scenario
                .when().placeOrder()
                .then().shouldSucceed()
                .and().order()
                    .hasStatus(OrderStatus.PLACED);
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void shouldCalculateBasePriceAsProductOfUnitPriceAndQuantity() {
        scenario
                .given().product()
                    .withUnitPrice(20.00)
                .when().placeOrder()
                    .withQuantity(5)
                .then().shouldSucceed()
                .and().order()
                    .hasBasePrice(100.00);
    }

    @TestTemplate
    @Channel(value = {ChannelType.API}, alsoForFirstRow = ChannelType.UI)
    @DataSource({"20.00", "5", "100.00"})
    @DataSource({"10.00", "3", "30.00"})
    @DataSource({"15.50", "4", "62.00"})
    @DataSource({"99.99", "1", "99.99"})
    void shouldPlaceOrderWithCorrectBasePriceParameterized(String unitPrice, String quantity, String basePrice) {
        scenario
                .given().product()
                    .withUnitPrice(unitPrice)
                .when().placeOrder()
                    .withQuantity(quantity)
                .then().shouldSucceed()
                .and().order()
                    .hasBasePrice(basePrice);
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void orderPrefixShouldBeORD() {
        scenario
                .when().placeOrder()
                .then().shouldSucceed()
                .and().order()
                    .hasOrderNumberPrefix("ORD-");
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void discountRateShouldBeAppliedForCoupon() {
        scenario
                .given().coupon()
                    .withCouponCode("SUMMER2025")
                    .withDiscountRate(0.15)
                .when().placeOrder()
                    .withCouponCode("SUMMER2025")
                .then().shouldSucceed()
                .and().order()
                    .hasAppliedCoupon("SUMMER2025")
                    .hasDiscountRate(0.15);
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void discountRateShouldBeNotAppliedWhenThereIsNoCoupon() {
        scenario
                .when().placeOrder()
                    .withCouponCode(null)
                .then().shouldSucceed()
                .and().order()
                    .hasAppliedCoupon(null)
                    .hasDiscountRate(0.00)
                    .hasDiscountAmount(0.00);
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void subtotalPriceShouldBeCalculatedAsTheBasePriceMinusDiscountAmountWhenWeHaveCoupon() {
        scenario
                .given().coupon()
                    .withDiscountRate(0.15)
                .and().product()
                    .withUnitPrice(20.00)
                .when().placeOrder()
                    .withCouponCode()
                    .withQuantity(5)
                .then().shouldSucceed()
                .and().order()
                    .hasAppliedCoupon()
                    .hasDiscountRate(0.15)
                    .hasBasePrice(100.00)
                    .hasDiscountAmount(15.00)
                    .hasSubtotalPrice(85.00);
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void subtotalPriceShouldBeSameAsBasePriceWhenNoCoupon() {
        scenario
                .given().product()
                    .withUnitPrice(20.00)
                .when().placeOrder()
                    .withQuantity(5)
                .then().shouldSucceed()
                .and().order()
                    .hasBasePrice(100.00)
                    .hasDiscountAmount(0.00)
                    .hasSubtotalPrice(100.00);
    }

    @TestTemplate
    @Channel(value = {ChannelType.API}, alsoForFirstRow = ChannelType.UI)
    @DataSource({"UK", "0.09"})
    @DataSource({"US", "0.20"})
    void correctTaxRateShouldBeUsedBasedOnCountry(String country, String taxRate) {
        scenario
                .given().country()
                    .withCode(country)
                    .withTaxRate(taxRate)
                .when().placeOrder()
                    .withCountry(country)
                .then().shouldSucceed()
                .and().order()
                    .hasTaxRate(taxRate);
    }

    @TestTemplate
    @Channel(value = {ChannelType.API}, alsoForFirstRow = ChannelType.UI)
    @DataSource({"UK", "0.09", "50.00", "4.50", "54.50"})
    @DataSource({"US", "0.20", "100.00", "20.00", "120.00"})
    void totalPriceShouldBeSubtotalPricePlusTaxAmount(String country, String taxRate, String subtotalPrice, String expectedTaxAmount, String expectedTotalPrice) {
        scenario
                .given().country()
                    .withCode(country)
                    .withTaxRate(taxRate)
                .and().product()
                    .withUnitPrice(subtotalPrice)
                .when().placeOrder()
                    .withCountry(country)
                    .withQuantity(1)
                .then().shouldSucceed()
                .and().order()
                    .hasTaxRate(taxRate)
                    .hasSubtotalPrice(subtotalPrice)
                    .hasTaxAmount(expectedTaxAmount)
                    .hasTotalPrice(expectedTotalPrice);
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void couponUsageCountHasBeenIncrementedAfterItsBeenUsed() {
        scenario
                .given().coupon()
                    .withCouponCode("SUMMER2025")
                .when().placeOrder()
                    .withCouponCode("SUMMER2025")
                .then().shouldSucceed()
                .and().coupon("SUMMER2025")
                    .hasUsedCount(1);
    }

    @TestTemplate
    @Channel(ChannelType.API)
    void orderTotalShouldIncludeTax() {
        scenario
                .given().country()
                    .withCode("DE")
                    .withTaxRate("0.19")
                .when().placeOrder()
                    .withCountry("DE")
                .then().shouldSucceed()
                .and().order()
                    .hasSubtotalPrice(20.00)
                    .hasTaxRate(0.19)
                    .hasTotalPrice(23.80);
    }

    @TestTemplate
    @Channel(ChannelType.API)
    void orderTotalShouldReflectCouponDiscount() {
        scenario
                .given().coupon()
                    .withCouponCode("DISC10")
                    .withDiscountRate(0.10)
                .when().placeOrder()
                    .withCouponCode("DISC10")
                .then().shouldSucceed()
                .and().order()
                    .hasSubtotalPrice(18.00)
                    .hasDiscountRate(0.10)
                    .hasAppliedCoupon("DISC10")
                    .hasTotalPrice(19.26);
    }

    @TestTemplate
    @Channel(ChannelType.API)
    void orderTotalShouldApplyCouponDiscountAndTax() {
        scenario
                .given().coupon()
                    .withCouponCode("COMBO10")
                    .withDiscountRate(0.10)
                .and().country()
                    .withCode("GB")
                    .withTaxRate("0.20")
                .when().placeOrder()
                    .withCountry("GB")
                    .withCouponCode("COMBO10")
                .then().shouldSucceed()
                .and().order()
                    .hasSubtotalPrice(18.00)
                    .hasDiscountRate(0.10)
                    .hasTaxRate(0.20)
                    .hasAppliedCoupon("COMBO10")
                    .hasTotalPrice(21.60);
    }
}

File: system-test/dotnet/SystemTests/Latest/AcceptanceTests/PlaceOrderPositiveTest.cs
using SystemTests.Latest.AcceptanceTests.Base;
using Dsl.Core.UseCase;
using Driver.Port.Dtos;
using Dsl.Port.Then.Steps;
using Optivem.Testing;

namespace SystemTests.Latest.AcceptanceTests;

public class PlaceOrderPositiveTest : BaseAcceptanceTest
{
    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task ShouldBeAbleToPlaceOrderForValidInput(Channel channel)
    {
        await Scenario(channel)
            .Given().Product().WithSku("ABC").WithUnitPrice(20.00m)
            .And().Country().WithCode("US").WithTaxRate(0.10m)
            .When().PlaceOrder().WithSku("ABC").WithQuantity(5).WithCountry("US")
            .Then().ShouldSucceed();
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task OrderStatusShouldBePlacedAfterPlacingOrder(Channel channel)
    {
        await Scenario(channel)
            .When().PlaceOrder()
            .Then().ShouldSucceed()
            .And().Order()
            .HasStatus(OrderStatus.Placed);
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task ShouldCalculateBasePriceAsProductOfUnitPriceAndQuantity(Channel channel)
    {
        await Scenario(channel)
            .Given().Product().WithUnitPrice(20.00m)
            .When().PlaceOrder().WithQuantity(5)
            .Then().ShouldSucceed()
            .And().Order()
            .HasBasePrice(100.00m);
    }

    [Theory]
    [ChannelData(ChannelType.API, AlsoForFirstRow = new[] { ChannelType.UI })]
    [ChannelInlineData("20.00", 5, "100.00")]
    [ChannelInlineData("10.00", 3, "30.00")]
    [ChannelInlineData("15.50", 4, "62.00")]
    [ChannelInlineData("99.99", 1, "99.99")]
    public async Task ShouldPlaceOrderWithCorrectBasePriceParameterized(Channel channel, string unitPrice, int quantity, string basePrice)
    {
        await Scenario(channel)
            .Given().Product().WithUnitPrice(unitPrice)
            .When().PlaceOrder().WithQuantity(quantity)
            .Then().ShouldSucceed()
            .And().Order()
            .HasBasePrice(basePrice);
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task OrderPrefixShouldBeORD(Channel channel)
    {
        await Scenario(channel)
            .When().PlaceOrder()
            .Then().ShouldSucceed()
            .And().Order()
            .HasOrderNumberPrefix("ORD-");
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task DiscountRateShouldBeAppliedForCoupon(Channel channel)
    {
        await Scenario(channel)
            .Given().Coupon().WithCouponCode("SUMMER2025").WithDiscountRate(0.15m)
            .When().PlaceOrder().WithCouponCode("SUMMER2025")
            .Then().ShouldSucceed()
            .And().Order()
            .HasAppliedCoupon("SUMMER2025")
            .HasDiscountRate(0.15m);
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task DiscountRateShouldBeNotAppliedWhenThereIsNoCoupon(Channel channel)
    {
        await Scenario(channel)
            .When().PlaceOrder().WithCouponCode(null)
            .Then().ShouldSucceed()
            .And().Order()
            .HasAppliedCoupon(null!)
            .HasDiscountRate(0.00m)
            .HasDiscountAmount(0.00m);
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task SubtotalPriceShouldBeCalculatedAsTheBasePriceMinusDiscountAmountWhenWeHaveCoupon(Channel channel)
    {
        await Scenario(channel)
            .Given().Coupon().WithDiscountRate(0.15m)
            .And().Product().WithUnitPrice(20.00m)
            .When().PlaceOrder().WithCouponCode().WithQuantity(5)
            .Then().ShouldSucceed()
            .And().Order()
            .HasAppliedCoupon()
            .HasDiscountRate(0.15m)
            .HasBasePrice(100.00m)
            .HasDiscountAmount(15.00m)
            .HasSubtotalPrice(85.00m);
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task SubtotalPriceShouldBeSameAsBasePriceWhenNoCoupon(Channel channel)
    {
        await Scenario(channel)
            .Given().Product().WithUnitPrice(20.00m)
            .When().PlaceOrder().WithQuantity(5)
            .Then().ShouldSucceed()
            .And().Order()
            .HasBasePrice(100.00m)
            .HasDiscountAmount(0.00m)
            .HasSubtotalPrice(100.00m);
    }

    [Theory]
    [ChannelData(ChannelType.API, AlsoForFirstRow = new[] { ChannelType.UI })]
    [ChannelInlineData("UK", 0.09)]
    [ChannelInlineData("US", 0.20)]
    public async Task CorrectTaxRateShouldBeUsedBasedOnCountry(Channel channel, string country, double taxRate)
    {
        await Scenario(channel)
            .Given().Country().WithCode(country).WithTaxRate((decimal)taxRate)
            .When().PlaceOrder().WithCountry(country)
            .Then().ShouldSucceed()
            .And().Order()
            .HasTaxRate((decimal)taxRate);
    }

    [Theory]
    [ChannelData(ChannelType.API, AlsoForFirstRow = new[] { ChannelType.UI })]
    [ChannelInlineData("UK", 0.09, "50.00", "4.50", "54.50")]
    [ChannelInlineData("US", 0.20, "100.00", "20.00", "120.00")]
    public async Task TotalPriceShouldBeSubtotalPricePlusTaxAmount(Channel channel, string country, double taxRate, string subtotalPrice, string expectedTaxAmount, string expectedTotalPrice)
    {
        await Scenario(channel)
            .Given().Country().WithCode(country).WithTaxRate((decimal)taxRate)
            .And().Product().WithUnitPrice(subtotalPrice)
            .When().PlaceOrder().WithCountry(country).WithQuantity(1)
            .Then().ShouldSucceed()
            .And().Order()
            .HasTaxRate((decimal)taxRate)
            .HasSubtotalPrice(subtotalPrice)
            .HasTaxAmount(expectedTaxAmount)
            .HasTotalPrice(expectedTotalPrice);
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task CouponUsageCountHasBeenIncrementedAfterItsBeenUsed(Channel channel)
    {
        await Scenario(channel)
            .Given().Coupon().WithCouponCode("SUMMER2025")
            .When().PlaceOrder().WithCouponCode("SUMMER2025")
            .Then().ShouldSucceed()
            .And().Coupon("SUMMER2025")
            .HasUsedCount(1);
    }

    [Theory]
    [ChannelData(ChannelType.API)]
    public async Task OrderTotalShouldIncludeTax(Channel channel)
    {
        await Scenario(channel)
            .Given().Country().WithCode("DE").WithTaxRate(0.19m)
            .When().PlaceOrder().WithCountry("DE")
            .Then().ShouldSucceed()
            .And().Order()
            .HasSubtotalPrice(20.00m)
            .HasTaxRate(0.19m)
            .HasTotalPrice(23.80m);
    }

    [Theory]
    [ChannelData(ChannelType.API)]
    public async Task OrderTotalShouldReflectCouponDiscount(Channel channel)
    {
        await Scenario(channel)
            .Given().Coupon().WithCouponCode("DISC10").WithDiscountRate(0.10m)
            .When().PlaceOrder().WithCouponCode("DISC10")
            .Then().ShouldSucceed()
            .And().Order()
            .HasSubtotalPrice(18.00m)
            .HasDiscountRate(0.10m)
            .HasAppliedCoupon("DISC10")
            .HasTotalPrice(19.26m);
    }

    [Theory]
    [ChannelData(ChannelType.API)]
    public async Task OrderTotalShouldApplyCouponDiscountAndTax(Channel channel)
    {
        await Scenario(channel)
            .Given().Coupon().WithCouponCode("COMBO10").WithDiscountRate(0.10m)
            .And().Country().WithCode("GB").WithTaxRate(0.20m)
            .When().PlaceOrder().WithCountry("GB").WithCouponCode("COMBO10")
            .Then().ShouldSucceed()
            .And().Order()
            .HasSubtotalPrice(18.00m)
            .HasDiscountRate(0.10m)
            .HasTaxRate(0.20m)
            .HasAppliedCoupon("COMBO10")
            .HasTotalPrice(21.60m);
    }
}
